source/log.c

```diff
@@ -0,0 +1,240 @@
+    thread_local_log_context->running = 1;
+    thread_local_log_context->max_message_len = max_message_len;
+    thread_local_log_context->alloc = alloc;
+    int ret = aws_memory_pool_init(&thread_local_log_context->message_pool, alloc, sizeof(struct aws_log_message) + max_message_len, memory_pool_message_count);
+    if (ret) {
+        aws_raise_error(AWS_ERROR_OOM);
+        return AWS_OP_ERR;
+    }
+    AWS_DOUBLY_LIST_INIT(thread_local_log_context);
+
+    /* Insert thread local log info struct onto the global doubly linked list used by
+    the logging thread. */
+    aws_mutex_lock(&global_log_list_mutex);
+    if (!global_log_list) {
+        global_log_list = thread_local_log_context;
+    } else {
+        AWS_DOUBLY_LIST_INSERT_BEFORE(global_log_list, thread_local_log_context);
+    }
+    aws_mutex_unlock(&global_log_list_mutex);
+
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_log_clean_up() {
+    if (aws_atomic_get(&global_log_thread_count) == 0) {
+        aws_raise_error(AWS_ERROR_LOG_UNINITIALIZED);
+        return AWS_OP_ERR;
+    }
+
+    struct aws_log_context *context;
+    do {
+        context = thread_local_log_context;
+    } while (!aws_atomic_compare_exchange_ptr((void **)&thread_local_log_context, context, NULL));
+
+    context->running = 0;
+
+    /* Perform final cleanup if this is the last log thread. */
+    aws_mutex_lock(&global_log_list_mutex);
+    if (aws_atomic_add(&global_log_thread_count, -1) == 1) {
+        struct aws_log_context *log_list = global_log_list;
+        global_log_list = NULL;
+        aws_mutex_unlock(&global_log_list_mutex);
+
+        /* Make sure all the other log threads have been joined and remove dead threads. */
+        struct aws_log_context *sentinel = log_list;
+        do {
+            struct aws_log_context *next = log_list->next;
+
+            /* Remove dead threads. Log any leftover messages. Cleanup leftover message pool memory. */
+            if (!log_list->running) {
+                struct aws_log_message *delete_list = log_list->message_list;
+                while (delete_list) {
+                    struct aws_log_message *next0 = delete_list->next;
+                    char *msg_data = (char*)(delete_list + 1);
+                    global_log_report_callback(msg_data);
+                    aws_memory_pool_release(&context->message_pool, delete_list);
+                    delete_list = next0;
+                }
+
+                delete_list = log_list->delete_list;
+                while (delete_list) {
+                    struct aws_log_message *next0 = delete_list->next;
+                    aws_memory_pool_release(&context->message_pool, delete_list);
+                    delete_list = next0;
+                }
+
+                aws_memory_pool_clean_up(&log_list->message_pool);
+            } else {
+                aws_raise_error(AWS_ERROR_LOG_IMPROPER_CLEAN_UP);
+                return AWS_OP_ERR;

```

Comment:
we may just want to assert() in the _clean_up() function, instead of returning error codes.
Especially this far into the function, where no we've detached the `global_log_list`, there's no recovery possible if one of these aws_log_contexts is still running

the thread-exit functions aren't checking these return codes anyway, so no one would notice if things went terribly off the rails, which is why I think asserts might be preferable to return codes here